### PR TITLE
fix gevent thread stuck (#458)

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Solve #458

This now works
```python
from src.taipy.gui import Gui, notify

import time

sleep_value = 1

md = """
<|{sleep_value}|number|>

<|Wait|button|on_action=wait|>
"""

def wait(state):
    notify(state,'i', 'First')
    time.sleep(state.sleep_value)
    notify(state,'i', 'Last')
    
Gui(md).run()
```

### CODE REVIEW

File: src/taipy/gui/gui.py [hunk 1]
```diff
@@ -25,6 +25,11 @@
 from types import FrameType
 from urllib.parse import urlparse
 
+if util.find_spec("gevent"):
```
Comment (FredLL-Avaiga) [inline src/taipy/gui/gui.py]: What happens if async_mode is set to something else than gevent ?

File: src/taipy/gui/gui.py [hunk 2]
```diff
@@ -1499,15 +1497,14 @@ def run(
...
                 Valid values are:</br>
 
                 - "threading": Use the Flask Development Server. This allows the application to use
                   the Flask reloader (the *use_reloader* option) and Debug mode (the *debug* option).
                 - "eventlet": Use eventlet server.
                 - "gevent": Use gevent server.
-                - "gevent_uwsgi": Use uwsgi server.
-
+                </br>The default value is "threading"</br>
                 If this argument is not set, Taipy uses, in that order: `"eventlet"`, `"gevent_uwsgi"`,
```
Comment (FredLL-Avaiga) [inline src/taipy/gui/gui.py]: this line needs to be changed too ?

File: src/taipy/gui/server.py [hunk 3]
```diff
@@ -180,7 +181,20 @@ def _is_port_open(self, host, port) -> bool:
         sock.close()
         return result == 0
 
+    def _apply_patch(self):
```
Comment (FredLL-Avaiga) [inline src/taipy/gui/server.py:208]: Nice ! 
A way to detect if patching has been done

**Review by FredLL-Avaiga — APPROVED** [review]

File: src/taipy/gui/server.py (hunk 3 above)
Comment (dinhlongviolin1) [reply src/taipy/gui/server.py:208]: Thanks!

File: src/taipy/gui/gui.py (hunk 2 above)
Comment (dinhlongviolin1) [reply src/taipy/gui/gui.py]: yes imo. gevent and eventlet are what will mainly be used

File: src/taipy/gui/server.py [hunk 4]
```diff
@@ -193,6 +205,18 @@ def _is_port_open(self, host, port) -> bool:
...
+        if self._get_async_mode() == "gevent" and util.find_spec("gevent"):
+            from gevent import monkey
+
+            if not monkey.is_anything_patched():
+                monkey.patch_all(ssl=False, thread=False)
+        if self._get_async_mode() == "eventlet" and util.find_spec("eventlet"):
```
Comment (FabienLelaquais) [inline src/taipy/gui/server.py:214]: That would be an 'else', right?

**Review by FabienLelaquais — APPROVED** [review]
Neat!!

File: src/taipy/gui/server.py (hunk 4 above)
Comment (dinhlongviolin1) [reply src/taipy/gui/server.py:214]: it is definitely and 'and'. Both conditions need to be met before any furhter execution

**Pull request merged**